### PR TITLE
'Solution'

### PR DESCRIPTION
**Pull request by tarasulo**

### CODE REVIEW

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 1]
```diff
@@ -3,48 +3,141 @@
...
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        if (first == null) {
+            first = new Node(value, null, null);
+            last = first;
+            size = 1;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: write size++
after if-else
and then remove them from lines 27 and 32

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 2]
```diff
@@ -3,48 +3,141 @@
...
+            throw new IndexOutOfBoundsException("The index is out of range: " + index);
+        }
+        if (index == size) {
+            add(value);
+        } else {
+            addValue(value);
+        }
+    }
 
+    private void addValue(T value) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]:
```suggestion
    private void addValueAtZeroPosition(T value) {
```

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 3]
```diff
@@ -3,48 +3,141 @@
...
+            Node newLast = new Node(value, last, null);
+            last.next = newLast;
+            last = newLast;
+            size++;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: where we add element at the special position? for example 4 or 7?

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 4]
```diff
@@ -3,48 +3,141 @@
...
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getElem(index).item;
+    }
+
+    public Node<T> getElem(int index) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]:
```suggestion
    private Node<T> getElement(int index) {
```

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 5]
```diff
@@ -3,48 +3,141 @@
...
+        } else {
+            previous.next = next;
+        }
+        if (next == null) {
+            last = previous;
+        } else {
+            next.prev = previous;
+        }
+        T removedData = removeElem.item;
+        removeElem.item = null;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]:
```suggestion
        removeElem = null;
```

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 6]
```diff
@@ -3,48 +3,141 @@
...
+        getElem(index).item = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+        Node<T> removeElem = getElem(index);
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]:
```suggestion
        Node<T> removeElement = getElem(index);
```

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 7]
```diff
@@ -3,48 +3,141 @@
...
+        T removedData = removeElem.item;
+        removeElem.item = null;
+        size--;
+        return removedData;
     }
 
     @Override
     public T remove(T t) {
+        for (int i = 0; i < size; i++) {
+            if (getElem(i).item.equals(t)) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: write: Node <T> node = getElem(i);
and then use `node`

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 8]
```diff
@@ -3,48 +3,141 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> first;
+    private Node<T> last;
+
+    private static class Node<T> {
+        T item;
```
[the remark this answers is not in the document]
Comment (tarasulo) [reply src/main/java/core/basesyntax/MyLinkedList.java]: done

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 1 above)
Comment (tarasulo) [reply src/main/java/core/basesyntax/MyLinkedList.java]: done

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 2 above)
Comment (tarasulo) [reply src/main/java/core/basesyntax/MyLinkedList.java]: done

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 4 above)
Comment (tarasulo) [reply src/main/java/core/basesyntax/MyLinkedList.java]: done

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 5 above)
Comment (tarasulo) [reply src/main/java/core/basesyntax/MyLinkedList.java]: done

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 6 above)
Comment (tarasulo) [reply src/main/java/core/basesyntax/MyLinkedList.java]: done

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 7 above)
Comment (tarasulo) [reply src/main/java/core/basesyntax/MyLinkedList.java]: done

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 9]
```diff
@@ -2,15 +2,17 @@
...
 
     private static class Node<T> {
-        T item;
-        Node next;
-        Node prev;
+        private T item;
+        private Node next;
+        private Node prev;
 
         Node(T item, Node prev, Node next) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: add access modifier

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 10]
```diff
@@ -41,20 +42,16 @@ public void add(T value, int index) {
...
-    private void addValue(T value) {
-        Node<T> firstNode = first;
-        Node<T> newNode = new Node<>(value, null, firstNode);
-        first = newNode;
-        if (firstNode == null) {
-            last = newNode;
-        } else {
-            firstNode.prev = newNode;
+            Node<T> oldNode = getElement(index);
+            Node<T> newNode = new Node(oldNode.prev, (Node) value, oldNode);
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: Node<T> newNode = new Node(value, oldNode.prev, oldNode);

Comment (tarasulo) [reply src/main/java/core/basesyntax/MyLinkedList.java]: done

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 9 above)
Comment (tarasulo) [reply src/main/java/core/basesyntax/MyLinkedList.java]: done

**Pull request closed without merging**